### ISSUE TITLE
How to specify a CPU time limit like qpOASES

### ISSUE THREAD

**Issue by zhilinxiong** [Simple-Robotics/proxsuite#226]
Thanks for your great work. I use the library to solve MPC problem for robot control, it is easier to use and more stable than qpOASES.
However, when the robot is perturbed, the QP solve time increase significantly (from 1~2 ms to above 10ms, and sometimes even over several seconds), which is unacceptable for my realtime applications. I want to know whether it supports to limit the cpu time or some other suggestions for limit the iterations.

**Comment by jcarpent** [comment]
@zhilinxiong Thanks for raising this issue. Could you provide an example in Python or C++ to replicate the behavior you have observed? We are ready to fix the issue if any.
We will add the time limit shortly.
@Bambade Could you handle this task?

**Comment by zhilinxiong** [comment]
> @zhilinxiong Thanks for raising this issue. Could you provide an example in Python or C++ to replicate the behavior you have observed? We are ready to fix the issue if any. We will add the time limit shortly. @Bambade Could you handle this task?

Thanks for your quick response. I'm really sorry, it's hard for me to provide an independent example right now since it contains the simulation and many other library to make the MPC running.
Could you give some suggestions about how to limit the solving time approximately?

**Comment by jcarpent** [comment]
Do you use the options for warm starting the problem from previous solutions already?

**Comment by zhilinxiong** [comment]
> Do you use the options for warm starting the problem from previous solutions already?

Here is my setting:
`
  /* proxsuite solver */
  // Timer proxsuite_timer;
  double eps_abs = 1e-9;
  proxsuite::proxqp::dense::isize dim = qp_dim, n_eq = 0, n_in = constraint_dim;
  proxsuite::proxqp::dense::QP<double> qp(dim, n_eq, n_in);
  qp.settings.eps_abs = eps_abs;
  // qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::NO_INITIAL_GUESS;
  qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::WARM_START_WITH_PREVIOUS_RESULT;
  qp.settings.verbose = false;
  // qp.settings.compute_timings = true;
  bool use_hotstart = last_problem_success_;
  if (use_hotstart) {
    qp.update(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
  } else {
    qp.init(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
  }
  qp.solve();
  if (qp.results.info.status == proxsuite::proxqp::QPSolverOutput::PROXQP_SOLVED) {
    Eigen::VectorXd qp_solution_reduced = qp.results.x;
    // recovery optimal solution from the reduced problem
    int buffer_index = 0;
    for (int leg = 0; leg < LEG_NUM; ++leg) {
      if (contact_sequence(leg, 0) == 0) {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM).setZero();
      } else {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM) = qp_solution_reduced.segment<WRENCH_DIM>(buffer_index * WRENCH_DIM);
        ++buffer_index;
      }
    }

    last_problem_success_ = true;
  }
`

**Comment by zhilinxiong** [comment]
> Do you use the options for warm starting the problem from previous solutions already?

```
  /* proxsuite solver */
  // Timer proxsuite_timer;
  double eps_abs = 1e-9;
  proxsuite::proxqp::dense::isize dim = qp_dim, n_eq = 0, n_in = constraint_dim;
  proxsuite::proxqp::dense::QP<double> qp(dim, n_eq, n_in);
  qp.settings.eps_abs = eps_abs;
  // qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::NO_INITIAL_GUESS;
  qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::WARM_START_WITH_PREVIOUS_RESULT;
  qp.settings.verbose = false;
  // qp.settings.compute_timings = true;
  bool use_hotstart = last_problem_success_;
  if (use_hotstart) {
    qp.update(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
  } else {
    qp.init(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
  }
  qp.solve();
  if (qp.results.info.status == proxsuite::proxqp::QPSolverOutput::PROXQP_SOLVED) {
    Eigen::VectorXd qp_solution_reduced = qp.results.x;
    // recovery optimal solution from the reduced problem
    int buffer_index = 0;
    for (int leg = 0; leg < LEG_NUM; ++leg) {
      if (contact_sequence(leg, 0) == 0) {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM).setZero();
      } else {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM) = qp_solution_reduced.segment<WRENCH_DIM>(buffer_index * WRENCH_DIM);
        ++buffer_index;
      }
    }

    last_problem_success_ = true;
  }
```

**Comment by jcarpent** [comment]
I would make this slight change for better efficiency. I have also added comments.
Importantly, you should create only once for all the qp problem, and update it according to the new measurements in an MPC fashion.
```cpp
  /* proxsuite solver */
  // Timer proxsuite_timer;
  double eps_abs = 1e-9; // You should increase this value. Your problem might be hard
  proxsuite::proxqp::dense::isize dim = qp_dim, n_eq = 0, n_in = constraint_dim;
  proxsuite::proxqp::dense::QP<double> qp(dim, n_eq, n_in); // this should be created only once. Otherwise, you are not taking into account the previous solutions.
  qp.settings.eps_abs = eps_abs;
  qp.settings.verbose = false;
  // qp.settings.compute_timings = true;
  bool use_hotstart = last_problem_success_;
  if (use_hotstart) {
    qp.update(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
    qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::WARM_START_WITH_PREVIOUS_RESULT;
  } else {
    qp.init(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
     qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::NO_INITIAL_GUESS;
  }
  qp.solve();
  if (qp.results.info.status == proxsuite::proxqp::QPSolverOutput::PROXQP_SOLVED) {
    const Eigen::VectorXd & qp_solution_reduced = qp.results.x;
    // recovery optimal solution from the reduced problem
    int buffer_index = 0;
    for (int leg = 0; leg < LEG_NUM; ++leg) {
      if (contact_sequence(leg, 0) == 0) {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM).setZero();
      } else {
        optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM) = qp_solution_reduced.segment<WRENCH_DIM>(buffer_index * WRENCH_DIM);
        ++buffer_index;
      }
    }

    last_problem_success_ = true;
  }
```

**Comment by zhilinxiong** [comment]
> I would make this slight change for better efficiency. I have also added comments. Importantly, you should create only once for all the qp problem, and update it according to the new measurements in an MPC fashion.
> 
> ```c++
>   /* proxsuite solver */
>   // Timer proxsuite_timer;
>   double eps_abs = 1e-9; // You should increase this value. Your problem might be hard
>   proxsuite::proxqp::dense::isize dim = qp_dim, n_eq = 0, n_in = constraint_dim;
>   proxsuite::proxqp::dense::QP<double> qp(dim, n_eq, n_in); // this should be created only once. Otherwise, you are not taking into account the previous solutions.
>   qp.settings.eps_abs = eps_abs;
>   qp.settings.verbose = false;
>   // qp.settings.compute_timings = true;
>   bool use_hotstart = last_problem_success_;
>   if (use_hotstart) {
>     qp.update(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
>     qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::WARM_START_WITH_PREVIOUS_RESULT;
>   } else {
>     qp.init(hessian_mat, gradient_vec, std::nullopt, std::nullopt, constraint_mat, std::nullopt, constraint_ub_vec);
>      qp.settings.initial_guess = proxsuite::proxqp::InitialGuessStatus::NO_INITIAL_GUESS;
>   }
>   qp.solve();
>   if (qp.results.info.status == proxsuite::proxqp::QPSolverOutput::PROXQP_SOLVED) {
>     const Eigen::VectorXd & qp_solution_reduced = qp.results.x;
>     // recovery optimal solution from the reduced problem
>     int buffer_index = 0;
>     for (int leg = 0; leg < LEG_NUM; ++leg) {
>       if (contact_sequence(leg, 0) == 0) {
>         optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM).setZero();
>       } else {
>         optimal_control_.segment<WRENCH_DIM>(leg * WRENCH_DIM) = qp_solution_reduced.segment<WRENCH_DIM>(buffer_index * WRENCH_DIM);
>         ++buffer_index;
>       }
>     }
> 
>     last_problem_success_ = true;
>   }
> ```

Thanks again. I adjusted the code according to your comments, at the first time, the computation time decrease about 2x. However, the solver may core dump when restarted.
I use gdb to debug the problem, it shows qpmodel is invalid as

```
main_control: /opt/control/include/proxsuite/proxqp/dense/helpers.hpp:387: void proxsuite::proxqp::dense::setup(proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::proxqp::Settings<T>&, proxsuite::proxqp::dense::Model<T>&, proxsuite::proxqp::dense::Workspace<T>&, proxsuite::proxqp::Results<T>&, proxsuite::proxqp::dense::preconditioner::RuizEquilibration<T>&, proxsuite::proxqp::PreconditionerStatus) [with T = double; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, -1, 1>, 0, Eigen::OuterStride<> > >; typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::OuterStride<>; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, 1>, 0, Eigen::InnerStride<1> > >; typename Eigen::internal::conditional<const Eigen::Matrix<Type, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::InnerStride<1>]: Assertion `qpmodel.is_valid()' failed.
```

**Comment by jcarpent** [comment]
Before moving further, which flags do you use for compiling your code?

**Comment by jcarpent** [comment]
Second, do you change the size of your QP matrices along the MPC iterations?
If so, you should create a fresh problem.

**Comment by zhilinxiong** [comment]
> Before moving further, which flags do you use for compiling your code?

My compiling is
```
set(CMAKE_BUILD_TYPE "Release")
set(BUILD_WITH_SIMD_SUPPORT ON)
add_compile_options(-Wall -Wextra -Wpedantic -O3 "-march=native")

add_library(cmpc_controller SHARED
  convex_mpc.cpp
)

target_link_libraries(cmpc_controller
  qpOASES
  # proxsuite::proxsuite
  proxsuite::proxsuite-vectorized
)
```

**Comment by zhilinxiong** [comment]
> Second, do you change the size of your QP matrices along the MPC iterations? If so, you should create a fresh problem.

I can make sure for now the QP matrices size is invariant. In the future, I will check whether the size is changed and then resize the problem.

**Comment by jcarpent** [comment]
> > Before moving further, which flags do you use for compiling your code?
> 
> My compiling is
> 
> ```
> set(CMAKE_BUILD_TYPE "Release")
> set(BUILD_WITH_SIMD_SUPPORT ON)
> add_compile_options(-Wall -Wextra -Wpedantic -O3 "-march=native")
> 
> add_library(cmpc_controller SHARED
>   convex_mpc.cpp
> )
> 
> target_link_libraries(cmpc_controller
>   qpOASES
>   # proxsuite::proxsuite
>   proxsuite::proxsuite-vectorized
> )
> ```

If you see the assertion:
```cpp
main_control: /opt/control/include/proxsuite/proxqp/dense/helpers.hpp:387: void proxsuite::proxqp::dense::setup(proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::proxqp::Settings<T>&, proxsuite::proxqp::dense::Model<T>&, proxsuite::proxqp::dense::Workspace<T>&, proxsuite::proxqp::Results<T>&, proxsuite::proxqp::dense::preconditioner::RuizEquilibration<T>&, proxsuite::proxqp::PreconditionerStatus) [with T = double; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, -1, 1>, 0, Eigen::OuterStride<> > >; typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::OuterStride<>; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, 1>, 0, Eigen::InnerStride<1> > >; typename Eigen::internal::conditional<const Eigen::Matrix<Type, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::InnerStride<1>]: Assertion `qpmodel.is_valid()' failed.
```
This means that you are in Debug mode or you don't have the `-DNDEBUG` activated.
To move further, I can help you if you share your code (in private of course).

**Comment by zhilinxiong** [comment]
> > > Before moving further, which flags do you use for compiling your code?
> > 
> > 
> > My compiling is
> > ```
> > set(CMAKE_BUILD_TYPE "Release")
> > set(BUILD_WITH_SIMD_SUPPORT ON)
> > add_compile_options(-Wall -Wextra -Wpedantic -O3 "-march=native")
> > 
> > add_library(cmpc_controller SHARED
> >   convex_mpc.cpp
> > )
> > 
> > target_link_libraries(cmpc_controller
> >   qpOASES
> >   # proxsuite::proxsuite
> >   proxsuite::proxsuite-vectorized
> > )
> > ```
> 
> If you see the assertion:
> 
> ```c++
> main_control: /opt/control/include/proxsuite/proxqp/dense/helpers.hpp:387: void proxsuite::proxqp::dense::setup(proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > >, proxsuite::proxqp::Settings<T>&, proxsuite::proxqp::dense::Model<T>&, proxsuite::proxqp::dense::Workspace<T>&, proxsuite::proxqp::Results<T>&, proxsuite::proxqp::dense::preconditioner::RuizEquilibration<T>&, proxsuite::proxqp::PreconditionerStatus) [with T = double; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<LhsScalar, -1, -1, 1>, 0, typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type> > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, -1, 1>, 0, Eigen::OuterStride<> > >; typename Eigen::internal::conditional<const Eigen::Matrix<LhsScalar, -1, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::OuterStride<>; proxsuite::optional<Eigen::Ref<const Eigen::Matrix<Type, -1, 1> > > = std::optional<Eigen::Ref<const Eigen::Matrix<double, -1, 1>, 0, Eigen::InnerStride<1> > >; typename Eigen::internal::conditional<const Eigen::Matrix<Type, -1, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::InnerStride<1>]: Assertion `qpmodel.is_valid()' failed.
> ```
> 
> This means that you are in Debug mode or you don't have the `-DNDEBUG` activated. To move further, I can help you if you share your code (in private of course).

Yes, add `-NDEBUG` solve the `qpmodel.is_valid()` assertion problem. But another issue arises 
```
[main_control-1] terminate called after throwing an instance of 'std::bad_array_new_length'
[main_control-1]   what():  std::bad_array_new_length
```
And I print intermediary results in gdb mode, it shows below

![image](https://github.com/Simple-Robotics/proxsuite/assets/22167464/25a020bc-91f0-471f-b3fb-286212a664c9)

**Comment by jcarpent** [comment]
We reach the limit of the introspection I can do remotely. Please provide access to your code if you want help.

**Comment by zhilinxiong** [comment]
> We reach the limit of the introspection I can do remotely. Please provide access to your code if you want help.

The QP problem should be reset whenever the warm start is failed, just calling init is not enough. Thank you again for your kindness.
